openedx_proversity_reports/reports/last_page_accessed.py

```diff
@@ -0,0 +1,183 @@
+                )
+
+                for vertical in vertical_blocks:
+                    for problem in blocks.get_children(vertical):
+                        if problem.block_id == last_completed_child_position.block_key.block_id:
+                            parent_tree = get_parent_tree(blocks, problem)
+                            parent_tree_name = '-'.join(get_parent_display_names(blocks, parent_tree))
+                            problem_parent = blocks.get_parents(problem)
+                            vertical_block_id = problem_parent[0].block_id
+
+                user_data.append({
+                    'username': user.username,
+                    'last_time_accessed': str(last_completed_child_position.modified),
+                    'last_page_viewed': parent_tree_name,
+                    'block_id': last_completed_child_position.block_key.block_id,
+                    'vertical_block_id': vertical_block_id,
+                })
+
+        if user_data:
+            last_page_data[course_id] = user_data
+
+    return last_page_data
+
+
+def get_parent_tree(root_block, unit_block):
+    """
+    Util function to get the parent block tree of the provided unit_block.
+    """
+    vertical_parent = root_block.get_parents(unit_block)
+    sequential_parent = root_block.get_parents(vertical_parent[0])
+    chapter_parent = root_block.get_parents(sequential_parent[0])
+    parent_tree = [chapter_parent[0], sequential_parent[0], vertical_parent[0], unit_block]
+
+    return parent_tree
+
+
+def get_parent_display_names(root_block, parent_tree):
+    """
+    Gets the parent display names of the provided parent_tree.
+    """
+    display_names = []
+    for block in parent_tree:
+        display_name = root_block.get_xblock_field(block, 'display_name')
+        display_names.append(unicode(display_name))
+
+    return display_names
+
+
+def get_exit_count_data(last_page_data, course_list):
+    """
+    Returns a dict containing all units in the each provided course_list.
+    Also return how many users are in each unit according to last_page_data data.
+
+    {
+        course_id: [{
+            'page_title': Chapter-Sequential-Vertical name,
+            'vertical_id': Vertical block id,
+            'exit_count': Number of users in this unit
+        }]
+    }
+    """
+
+    course_unit_data = {}
+
+    for course_id in course_list:
+        try:
+            course_key = CourseKey.from_string(course_id)
+        except InvalidKeyError:
+            continue
+
+        # Getting all students enrolled in the course except staff users
+        enrolled_students = User.objects.filter(
+            courseenrollment__course_id=course_key,
+            courseenrollment__is_active=1,
+            is_staff=0,
+        )
+
+        if not enrolled_students:
+            continue
+
+        usage_key = get_modulestore().make_course_usage_key(course_key)
+        blocks = get_course_blocks(enrolled_students[0], usage_key)
```

Comment:
what happen if enrolled_students is empty, does this need a try except block ?

Reply:
Line 139 will catch it. @andrey-canon